app/src/main/kotlin/com/waz/zclient/settings/account/SettingsAccountViewModel.kt

```
@@ -1,61 +1,68 @@
-    private val mutableError = MutableLiveData<String>()
     private val mutableProfile = MutableLiveData<UserProfileItem>()
+    private val mutableNameUpdated = MutableLiveData<Boolean>().apply { setValue(false) }
+    private val mutableError = MutableLiveData<String>()
+
+    val profile: LiveData<UserProfileItem>
+        get() = mutableProfile
 
-    val loading: LiveData<Boolean>
-        get() = mutableLoading
+    val nameUpdated: LiveData<Boolean>
+        get() = mutableNameUpdated
 
     val error: LiveData<String>
         get() = mutableError
 
-    val profile: LiveData<UserProfileItem>
-        get() = mutableProfile
 
-    fun loadData() {
-        handleLoading(true)
-        getUserProfileUseCase(viewModelScope, Unit) { response ->
-            response.fold(::handleProfileError, ::handleProfileSuccess)
-        }
+    fun loadProfile() {
+        getUserProfileUseCase(scope = viewModelScope, params = Unit,
+            onSuccess = { user -> handleProfileSuccess(user) },
+            onError = { handleError(it) }
+        )
     }
 
-    private fun handleProfileError(failure: Failure) {
-        handleLoading(false)
-        when (failure) {
-            is HttpError ->
-                Log.e(javaClass.simpleName, "failed with errorCode: ${failure.errorCode} and errorMessage {${failure.errorMessage}")
-            else ->
-                Log.e(javaClass.simpleName, "Misc error scenario")
-        }
+    fun updateName(value: String) {
+        changeNameUseCase(scope = viewModelScope, params = ChangeNameParams(value),
+            onSuccess = { handleChangeNameSuccess() },
+            onError = { handleError(it) }
+        )
     }
 
     private fun handleProfileSuccess(user: User) {
-        handleLoading(false)
         mutableProfile.postValue(UserProfileItem(user))
     }
 
-    private fun handleLoading(isLoading: Boolean) {
-        mutableLoading.postValue(isLoading)
+    private fun handleError(throwable: Throwable) {
```

Should we tightly couple this to a Throwable? We have the sealed class for Failures to be handled and it's more flexible

---

I cannot agree more

---

Refactored to accomodate Failure's. Just need to propagate the error properly.